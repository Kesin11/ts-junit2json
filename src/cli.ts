#!/usr/bin/env node
import fs from 'fs'
import yargs from 'yargs'
import { hideBin } from 'yargs/helpers'
import { parse } from './index.js'

const argv = yargs(hideBin(process.argv))
  .usage('Usage: $0 [path]')
  .command('$0 <path>', 'Convert JUnit XML format to JSON')
  .options({
    p: { type: 'boolean', alias: 'pretty', describe: 'Output pretty JSON', default: false },
    f: { type: 'string', alias: 'filter-tags', describe: 'Filter XML tag names' },
  })
  .alias('p', 'pretty')
  .alias('f', 'filter-tags')
  .coerce('f', (filter?: string) => filter?.split(',') )
  .positional('path', {
    describe: 'JUnit XML path',
    type: 'string',
    demandOption: true,
  })
  .example('$0 -p -f system-out,system-err junit.xml', 'Output pretty JSON with filter <system-out> and <system-err> tags.')
  .strict()
  .parseSync()

const main = async () => {
  const xmlString = fs.readFileSync(argv.path, 'utf8').toString()

  const filterSet = new Set(argv.filterTags)
  const _replacer = (key: any, value: any) => {
    if (filterSet.has(key)) return undefined
    return value
  }

  const indent = (argv.pretty) ? 2 : 0
  const replacer = (argv.filterTags && argv.filterTags.length > 0) ? _replacer : undefined

  const output = await parse(xmlString)
  console.log(JSON.stringify(output, replacer, indent))
}
main()