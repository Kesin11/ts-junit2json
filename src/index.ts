import xml2js from 'xml2js'

export const parse = async (xmlString: xml2js.convertableToString) => {
  const result = await xml2js.parseStringPromise(xmlString, {
    attrValueProcessors: [xml2js.processors.parseNumbers]
  })
  const rawTestsuites = result['testsuites']
  const output = { ...rawTestsuites['$'] }

  const testsuiteList = []
  for (const rawTestsuite of rawTestsuites['testsuite']) {
    const testsuite = { ...rawTestsuite['$'] }

    const testcaseList = []
    for (const rawTestcase of rawTestsuite['testcase']) {
      const testcase = { ...rawTestcase['$'] }
      // TODO: キーの中身を見て$、_、配列であれば展開するように一般化できるはず
      const rawFailure = rawTestcase['failure']
      if (rawFailure) {
        const failureList = []
        for (const rawFailure of rawTestcase['failure']) {
          let failure = {}
          if (typeof rawFailure === 'object') {
            failure = { ...rawFailure['$'] }
            if (rawFailure['_']) {
              failure = { ...failure, body: rawFailure['_'] }
            }
          }
          // attributeが何も存在しない場合、$や_は存在せずにstring型でタグの中身が文字列として入っている
          else {
            failure = { body: rawFailure }
          }
        failureList.push(failure)
        }
        testcase['failure'] = failureList
      }

      testcaseList.push(testcase)
    }
    testsuite['testcase'] = testcaseList
    testsuiteList.push(testsuite)
  }
  output['testsuite'] = testsuiteList

  return output
}