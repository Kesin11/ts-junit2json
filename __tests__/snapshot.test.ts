import * as junit2json from '../src/index'
import * as fs from 'fs'
import * as path from 'path'

const fixturePath = (fixtureName: string) => {
  return path.join(__dirname, 'fixtures', fixtureName)
}

describe('parse snapshot', () => {
  it('all success xml', async () => {
    const xml = fs.readFileSync(fixturePath('jest-success.xml'))
    const parsed = await junit2json.parse(xml)

    expect(parsed).toMatchSnapshot()
  })

  it('jest failure xml', async () => {
    const xml = fs.readFileSync(fixturePath('jest-failure.xml'))
    const parsed = await junit2json.parse(xml)

    expect(parsed).toMatchSnapshot()
  })

  it('nunit failure xml', async () => {
    const xml = fs.readFileSync(fixturePath('nunit-failure.xml'))
    const parsed = await junit2json.parse(xml)

    expect(parsed).toMatchSnapshot()
  })
})
