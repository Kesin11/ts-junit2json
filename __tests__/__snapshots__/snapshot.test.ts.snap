// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse snapshot all success xml 1`] = `
Object {
  "failures": 0,
  "name": "gcf_junit_xml_to_bq_dummy",
  "tests": 12,
  "testsuite": Array [
    Object {
      "errors": 0,
      "failures": 0,
      "name": "__tests__/gen_dummy_junit/dummy2.test.js",
      "skipped": 0,
      "testcase": Array [
        Object {
          "classname": "dummy2 Always success tests should be wait 0-2sec",
          "name": "dummy2 Always success tests should be wait 0-2sec",
          "time": 0.045,
        },
        Object {
          "classname": "dummy2 Always success tests should be wait 1-3sec",
          "name": "dummy2 Always success tests should be wait 1-3sec",
          "time": 1.992,
        },
        Object {
          "classname": "dummy2 Randomly fail tests should be wait 0-1sec and fail 30%",
          "name": "dummy2 Randomly fail tests should be wait 0-1sec and fail 30%",
          "time": 0.062,
        },
        Object {
          "classname": "dummy2 Randomly fail tests should be wait 1-2sec and fail 10%",
          "name": "dummy2 Randomly fail tests should be wait 1-2sec and fail 10%",
          "time": 1.289,
        },
      ],
      "tests": 4,
      "time": 4.189,
      "timestamp": "2020-01-12T16:33:39",
    },
    Object {
      "errors": 0,
      "failures": 0,
      "name": "__tests__/gen_dummy_junit/dummy3.test.js",
      "skipped": 0,
      "testcase": Array [
        Object {
          "classname": "dummy3 Always success tests should be wait 0-2sec",
          "name": "dummy3 Always success tests should be wait 0-2sec",
          "time": 1.875,
        },
        Object {
          "classname": "dummy3 Always success tests should be wait 1-3sec",
          "name": "dummy3 Always success tests should be wait 1-3sec",
          "time": 1.346,
        },
        Object {
          "classname": "dummy3 Randomly fail tests should be wait 0-1sec and fail 30%",
          "name": "dummy3 Randomly fail tests should be wait 0-1sec and fail 30%",
          "time": 0.11,
        },
        Object {
          "classname": "dummy3 Randomly fail tests should be wait 1-2sec and fail 20%",
          "name": "dummy3 Randomly fail tests should be wait 1-2sec and fail 20%",
          "time": 1.211,
        },
      ],
      "tests": 4,
      "time": 5.309,
      "timestamp": "2020-01-12T16:33:39",
    },
    Object {
      "errors": 0,
      "failures": 0,
      "name": "__tests__/gen_dummy_junit/dummy1.test.js",
      "skipped": 0,
      "testcase": Array [
        Object {
          "classname": "dummy1 Always success tests should be wait 0-2sec",
          "name": "dummy1 Always success tests should be wait 0-2sec",
          "time": 0.494,
        },
        Object {
          "classname": "dummy1 Always success tests should be wait 1-3sec",
          "name": "dummy1 Always success tests should be wait 1-3sec",
          "time": 1.466,
        },
        Object {
          "classname": "dummy1 Randomly fail tests should be wait 0-1sec and fail 50%",
          "name": "dummy1 Randomly fail tests should be wait 0-1sec and fail 50%",
          "time": 0.73,
        },
        Object {
          "classname": "dummy1 Randomly fail tests should be wait 1-2sec and fail 30%",
          "name": "dummy1 Randomly fail tests should be wait 1-2sec and fail 30%",
          "time": 1.952,
        },
      ],
      "tests": 4,
      "time": 5.406,
      "timestamp": "2020-01-12T16:33:39",
    },
  ],
  "time": 5.82,
}
`;

exports[`parse snapshot jest failure xml 1`] = `
Object {
  "failures": 2,
  "name": "gcf_junit_xml_to_bq_dummy",
  "tests": 12,
  "testsuite": Array [
    Object {
      "errors": 0,
      "failures": 1,
      "name": "__tests__/gen_dummy_junit/dummy1.test.js",
      "skipped": 0,
      "testcase": Array [
        Object {
          "classname": "dummy1 Always success tests should be wait 0-2sec",
          "name": "dummy1 Always success tests should be wait 0-2sec",
          "time": 0.414,
        },
        Object {
          "classname": "dummy1 Always success tests should be wait 1-3sec",
          "name": "dummy1 Always success tests should be wait 1-3sec",
          "time": 1.344,
        },
        Object {
          "classname": "dummy1 Randomly fail tests should be wait 0-1sec and fail 50%",
          "failure": Array [
            Object {
              "inner": "Error: expect(received).toBeGreaterThan(expected)

Expected: > 50
Received:   4.897277513425746
    at Object.it (/Users/kesin/github/gcf_junit_xml_to_bq/__tests__/gen_dummy_junit/dummy1.test.js:22:17)",
            },
          ],
          "name": "dummy1 Randomly fail tests should be wait 0-1sec and fail 50%",
          "time": 0.673,
        },
        Object {
          "classname": "dummy1 Randomly fail tests should be wait 1-2sec and fail 30%",
          "name": "dummy1 Randomly fail tests should be wait 1-2sec and fail 30%",
          "time": 1.604,
        },
      ],
      "tests": 4,
      "time": 4.772,
      "timestamp": "2020-01-12T16:33:13",
    },
    Object {
      "errors": 0,
      "failures": 1,
      "name": "__tests__/gen_dummy_junit/dummy3.test.js",
      "skipped": 0,
      "testcase": Array [
        Object {
          "classname": "dummy3 Always success tests should be wait 0-2sec",
          "name": "dummy3 Always success tests should be wait 0-2sec",
          "time": 1.328,
        },
        Object {
          "classname": "dummy3 Always success tests should be wait 1-3sec",
          "name": "dummy3 Always success tests should be wait 1-3sec",
          "time": 2.598,
        },
        Object {
          "classname": "dummy3 Randomly fail tests should be wait 0-1sec and fail 30%",
          "failure": Array [
            Object {
              "inner": "Error: expect(received).toBeGreaterThan(expected)

Expected: > 30
Received:   12.15901879426653
    at Object.it (/Users/kesin/github/gcf_junit_xml_to_bq/__tests__/gen_dummy_junit/dummy3.test.js:22:17)",
            },
          ],
          "name": "dummy3 Randomly fail tests should be wait 0-1sec and fail 30%",
          "time": 0.455,
        },
        Object {
          "classname": "dummy3 Randomly fail tests should be wait 1-2sec and fail 20%",
          "name": "dummy3 Randomly fail tests should be wait 1-2sec and fail 20%",
          "time": 1.228,
        },
      ],
      "tests": 4,
      "time": 6.372,
      "timestamp": "2020-01-12T16:33:13",
    },
    Object {
      "errors": 0,
      "failures": 0,
      "name": "__tests__/gen_dummy_junit/dummy2.test.js",
      "skipped": 0,
      "testcase": Array [
        Object {
          "classname": "dummy2 Always success tests should be wait 0-2sec",
          "name": "dummy2 Always success tests should be wait 0-2sec",
          "time": 1.937,
        },
        Object {
          "classname": "dummy2 Always success tests should be wait 1-3sec",
          "name": "dummy2 Always success tests should be wait 1-3sec",
          "time": 2.904,
        },
        Object {
          "classname": "dummy2 Randomly fail tests should be wait 0-1sec and fail 30%",
          "name": "dummy2 Randomly fail tests should be wait 0-1sec and fail 30%",
          "time": 0.105,
        },
        Object {
          "classname": "dummy2 Randomly fail tests should be wait 1-2sec and fail 10%",
          "name": "dummy2 Randomly fail tests should be wait 1-2sec and fail 10%",
          "time": 1.948,
        },
      ],
      "tests": 4,
      "time": 7.612,
      "timestamp": "2020-01-12T16:33:13",
    },
  ],
  "time": 8.018,
}
`;

exports[`parse snapshot nunit failure xml 1`] = `
Object {
  "failures": 1,
  "name": "Tutorial.Tests.dll",
  "tests": 4,
  "testsuite": Array [
    Object {
      "errors": 0,
      "failures": 1,
      "hostname": "executor://xunit/VsTestRunner2/netcoreapp",
      "name": "Tutorial.Tests.dll",
      "skipped": 0,
      "testcase": Array [
        Object {
          "classname": "Tutorial.Tests.PersonClassTest",
          "file": "/Users/kesin/github/csharp-tutorial/tests/Tutorial.Tests/bin/Debug/netcoreapp3.1/Tutorial.Tests.dll",
          "name": "Tutorial.Tests.PersonClassTest.PersonClass_ToString",
          "time": 0.0039775,
        },
        Object {
          "classname": "Tutorial.Tests.PersonTest",
          "failure": Array [
            Object {
              "inner": "   at Tutorial.Tests.PersonTest.Person_fname() in /Users/kesin/github/csharp-tutorial/tests/Tutorial.Tests/PersonTest.cs:line 23
",
              "message": "Assert.Equal() Failure
          ↓ (pos 0)
Expected: foge
Actual:   john
          ↑ (pos 0)",
              "type": "failure",
            },
          ],
          "file": "/Users/kesin/github/csharp-tutorial/tests/Tutorial.Tests/bin/Debug/netcoreapp3.1/Tutorial.Tests.dll",
          "name": "Tutorial.Tests.PersonTest.Person_fname",
          "time": 0.0063612,
        },
        Object {
          "classname": "Tutorial.Tests.PersonTest",
          "file": "/Users/kesin/github/csharp-tutorial/tests/Tutorial.Tests/bin/Debug/netcoreapp3.1/Tutorial.Tests.dll",
          "name": "Tutorial.Tests.PersonTest.Person_Name",
          "time": 0.0001106,
        },
        Object {
          "classname": "Tutorial.Tests.PersonTest",
          "file": "/Users/kesin/github/csharp-tutorial/tests/Tutorial.Tests/bin/Debug/netcoreapp3.1/Tutorial.Tests.dll",
          "name": "Tutorial.Tests.PersonTest.Person_lname",
          "time": 0.0001331,
        },
      ],
      "tests": 4,
      "time": 0.0105824,
      "timestamp": "2019-12-31T 09:33:25Z",
    },
  ],
  "time": 0.0105824,
}
`;
